### DISCUSSION TITLE
RFC: Add an option to not reset state on navigation

### DISCUSSION THREAD

**Discussion by remorses** [vercel/next.js#49749] [Ideas]
### Goals

The new app router resets layout and page state when navigating between pages unser a dynamic segments like `[lang]` or `[[...slug]]`.

Adding the option `componentKey = 'manual'` would allow the user to keep state between navigations of a `Layout` or `Template` component

## Use cases

1. A docs website with a sidebar layout can keep the scroll position of the sidebar after a navigation. The sidebar layout would be under the dynamic segment because it depends on slug parameters to do data fetching.  #48082
2. Use a root `Template` component for page transitions without remounting every page on every navigation (as an alternative to `Layout` that [does not rerender on navigation](https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#layouts))

### Background

The old `_app.jsx` behaviour was to keep state between navigations and to instead [manually pass `<Component key={path} {...pageProps} />` to force a state reset on navigation](https://nextjs.org/docs/pages/api-reference/functions/use-router#resetting-state-after-navigation).

The current layout key is applied here, this proposal would simply skip passing this key
https://github.com/vercel/next.js/blob/9028a169acb04c208844582866c7317dfc336580/packages/next/src/client/components/layout-router.tsx#L567

### Proposal

On layout and template files you can pass `componentKey = 'manual'` to preserve state between navigations

If the `Layout` is not under a dynamic segment (like `[lang]` or `[[...slug]]`) this option will have no effect because common layouts are not rerendered.

The `Template` file instead will rerender but will not change `key`, making it possible to use page transition animations without resetting state between navigations

```ts
// [slug]/layout.tsx

export const componentKey = 'manual'

// does not reset state on navigation
export default function Layout({ params: { slug } }) {
  return <div>{children}</div>
}

```

The possible values of `componentKey` are `manual` and `auto` (the default)

An alternative to this proposal would be to just remove the `key` prop and let users manually pass `key={slug}` to have the current behaviour (just like it was done with `_app.jsx`)

**Comment by joacub** [comment]
I also would extend this functionality to page.tsx to archive same behavior if desired.

the case of [lang] is one of the cases you don’t want page to lose the sate.

**Comment by jkuhs** [comment]
It would be nice to understand why the key is being set this way in the first place.

There was a change recently that fixed the exact same issue on search parameter changes: https://github.com/vercel/next.js/commit/25ba8a74b7544dfb6b30d1b67c47b9cb5360cb4e

@feedthejim Since you fixed that issue, can you maybe speak to the related issue of the key changing when dynamic segment parameters change? 

Conceptually, I'm not sure I see much of a difference between search parameters changing vs dynamic segment parameters changing. In both cases, the same React tree should remain in place. The only thing that should change is the context (accessible via `useContext`, `useParams`, `useSearchParams`, etc.).

Am I missing something?

**Comment by jkuhs** [comment]
Also possibly relevant is this PR by @timneutkens: https://github.com/vercel/next.js/commit/d32ee25bfbe92719778a37529d1c3cb5df767e39

He provided some good details in the description. Among other things, he mentions:

> The router cache key is based on the "static key" / "dynamic parameter value" in the tree.

I won't pretend I have a full grasp on the various use cases and implications of that cache key. I presume that it's needed in some other parts of the app. It would be great if @timneutkens could speak to that. 

Even if the "dynamic parameter value" is needed in the cache key for things to work in other places, I would think it would still be  reasonable to change the logic around the TemplateContext.Provider key value to avoid a complete remount of the dynamic segment.

**Reply by joacub** [reply]
If they change the key for the segment it will be fixed both sides. Don’t know why they didn’t do that.

Example key with no collisions:

key = /[lang]/[other]
key2 = /es/something

different keys no collisions.
That both keys match both so if you create a static route it should have a priority if not go next….

**Reply by jkuhs** [reply]
Right, as far as the issue we are seeing with the `<TemplateContext.Provider key="...">` is concerned, I completely agree that the "static key" should be used, like you described.

I'm assuming there is a different reason for why they include the "dynamic parameter value". Maybe to cache data which might change when the "dynamic parameter value" changes. If that's the case, then I would argue that separate cache keys should be used for each use case.

**Reply by joacub** [reply]
They included cause same routers has colisión and were matching different routes in server

**Comment by Tankenstein** [comment]
This would help me a lot – we're building a pretty standard intercepting route modal/detail page setup right now, but you should be able to navigate to other instances from within the modal. This is too complicated to handle with file-based routes, but there's currently no way to opt out of nextjs routing due to it setting the key, even with catch-all routes.

**Comment by remorses** [comment]
A good compromise would be not setting the key for layouts like `[param]/layout.jsx`, not sure how to implement something like this tho

**Comment by M7ilan** [comment]
I saw issue #48082 and I have the same issue here but with [parallel routes](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes), I also have created an issue #60037.

idk if there's a solution now but your feedback looks a good solution for this.

**Comment by cbratschi** [comment]
I ran into this issue a couple of months ago and failed to find this discussion at that time. Let me present first my use case and problems with finding a workaround.

Our app uses a /[lang]/[[...slug]] route to render pages from a CMS. Any used client component in either layout.tsx or page.tsx is fully reset during a page navigation e.g. from /en/a to /en/b. Most of the client components produce the exactly same DOM again but there are a few cases where some properties change i.e. a re-rendering would occur in normal React.

Therefore strange issues occur due to the lost state: network request are done on each page navigation and screen flickers occurs because some components render their content after the network requests are done. Images and other resources might be reloaded by the browser.

For instance we use an admin bar component from the CMS which gets the ID of the current page. This component internally calls fetch to check if the user is logged-in to the CMS (server-side cookie is checked). In an SPA this would only occur once in the SPA's lifetime but in this Next.js route it occurs again and again. In more complex components this will have more dramatic consequences and  has performance impacts on client/server side besides screen reflow issues.

For this use case it's not easy to write a workaround and in my opinion no workaround should be needed for such cases. Maintaining the React state is a fundamental feature and any side-effects should be clearly listed in the Next.js docs or even output to debug logs in dev mode.

My workaround would be (again, I don't want to implement this): move the state outside the component using a global stage handler, such as Zustand. Fork the original React component and re-implement it's business logic using the global state hooks. This should work but it would be quite some overhead and each and every client component would have to be re-worked to use the global state handler. For instance search fields in a navbar are affected too by the state loss.

So in my opinion Next.js should always keep the React state in (dynamic) routes. This is such an important feature of React server components. The current implementation is due to the different createRouterCacheKey(). This should be reviewed again to support client components keeping their state and making things a lot easier. A lot of people will run into this issue sooner or later.

I wrote a small demo showing the issue:

https://github.com/cbratschi/nextjs-issues/tree/react-lifecycle/src/app/%5Blang%5D/%5B%5B...slug%5D%5D

Links to related issues:

- https://github.com/vercel/next.js/issues/48082
- https://github.com/vercel/next.js/issues/49185
- https://github.com/vercel/next.js/issues/49553
- https://github.com/vercel/next.js/issues/58699

**Reply by jkuhs** [reply]
@cbratschi You might want to also chime in on https://github.com/vercel/next.js/discussions/50711 where one of the maintainers posted an answer just recently.

**Reply by cbratschi** [reply]
@jkuhs Thanks, I missed your excellent summary and will follow-up there too.

**Comment by mfal** [comment]
For all those who need a temporary solution. You can use [`@mittwald/react-tunnel`](https://www.npmjs.com/package/@mittwald/react-tunnel?activeTab=readme) to lift the state from the page into the layout. 

This means that no state is lost when navigating between pages with dynamic segments.

Also see this [Codesandbox](https://codesandbox.io/p/devbox/lnt4v8?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm0mv8s0e00063b5t3j4058tx%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm0mv8s0d00023b5tdc1pkh0m%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm0mv8s0d00043b5teho5a5st%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm0mv8s0e00053b5t1g4ubwn3%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm0mv8s0d00023b5tdc1pkh0m%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0mv8s0d00013b5txumui2ib%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cm0mv8s0d00023b5tdc1pkh0m%2522%252C%2522activeTabId%2522%253A%2522cm0mv8s0d00013b5txumui2ib%2522%257D%252C%2522cm0mv8s0e00053b5t1g4ubwn3%2522%253A%257B%2522id%2522%253A%2522cm0mv8s0e00053b5t1g4ubwn3%2522%252C%2522activeTabId%2522%253A%2522cm0nhl0uy01dh3b5t8kpvq9wn%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cm0mvr4h2008p3b5t8mtlcikq%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522cm0nhb8gr01803b5tmg9tsnpc%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522SETUP_TASKS%2522%252C%2522id%2522%253A%2522cm0nhl0uy01dh3b5t8kpvq9wn%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522VIRTUAL_MACHINE%2522%252C%2522id%2522%253A%2522cm0nhoyam01fb3b5t9gpqi0rm%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm0mv8s0d00043b5teho5a5st%2522%253A%257B%2522id%2522%253A%2522cm0mv8s0d00043b5teho5a5st%2522%252C%2522activeTabId%2522%253A%2522cm0mv8s0d00033b5tdifctgh5%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm0mv8s0d00033b5tdifctgh5%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522install%2522%252C%2522id%2522%253A%2522cm0nku0ur000u3b5t6z23llzp%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D).

```tsx
// layout.tsx
"use client";
import { PropsWithChildren } from "react";
import { TunnelExit, TunnelProvider } from "@mittwald/react-tunnel";

export default function Layout(props: PropsWithChildren) {
  const { children } = props;

  return (
    <>
      <h1>My App</h1>
      <TunnelProvider>
        {children}
        <TunnelExit />
      </TunnelProvider>
    </>
  );
}
```

```tsx
// [id]/page.tsx
"use client";
import { useParams } from "next/navigation";
import { TunnelEntry } from "@mittwald/react-tunnel";

export default function Page() {
  const postId = useParams().id;

  return (
    <TunnelEntry staticEntryId="BlogPage">
      <h2>Post #{postId}</h2>
      <textarea defaultValue="Change me – I will keep me state" />
    </TunnelEntry>
  );
}
```
